XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+| Title  | AsyncWrap API |
+|--------|---------------|
+| Author | @trevnorris   |
+| Status | DRAFT         |
+| Date   | 2016-04-05    |
+
+## Description
+
+Since its initial introduction with the `AsyncListener` API, `AsyncWrap` has
+slowly evolved to ensure a generalized API that would serve as a solid base for
+module authors who wished to add events to the life cycle of the event loop.
+The API presented below aims to accomplish this.
+
+## API
+
+### Overview
+
+```js
+// Standard way of requiring a module. Snake case follows core module
+// convention.
+const async_wrap = require('async_wrap');
+
+// List of asynchronous providers (e.g. CRYPTO or TCPWRAP). Can be used for
+// trace filtering of events.
+const providers = async_wrap.Providers;
+
+// init() is called during object construction. The object handle being
+// constructed can be accessed as "this". Because init() is called during
+// construction it is not safe to blindly inspect the constructed object at the
+// moment. Though we are currently investigating how to make this safe.
+function init(id, type, parentId) { }
+
+// pre() is called just before the handle's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWRAP), and should be called exactly 1
+// time for requests (e.g. FSREQWRAP).
+function pre(id) { }
```

Comment:
Please don't use the term "callback". That complicates the existing meaning. Which are the callbacks attached to the handle. It would be more clear if you simply used the term "handle" for both a handle and request. The only time I differentiate the two is when there are important technical differences.

> A _callback_ is a "registration" of a javascript function that is to be called asynchronously at some later point during program execution.

This is technically inaccurate and confusing. Callbacks are never "registered". They are simply used by handles to notify when work has been done (e.g. new connection was made). Callbacks also don't contain state, which is an important part of the existence of a handle.

`invocation_id` is no necessary to keep internally. We should expose the callback type to the user. e.g. TCP can call either an onread or onconnection callback, but from there it's easy to track that count themselves with little overhead. while it would require a more than trivial change to core to be done in an acceptable way.